### ISSUE TITLE
Support custom file type diagnostics

### ISSUE THREAD

**Issue by glyh** [rust-lang/rust-analyzer#17052]
For example, there is LALRPOP that is basically a DSL that mixes rust and custom constructs. It should be possible to let rust_analyzer analyze this. 

I think it would be good if there's some kind of API to let end user to implement this.

**Comment by Veykril** [comment]
That feels out of scope in my opinion. rust-analyzer is about rust, not about DSL languages that are like rust

**Comment by glyh** [comment]
If you look at them closely, some part of the codes are just rust code. Plus I'm not asking such support should be in the main repo. I'm just saying if there's some kind of API to let downstream DSL creators to integrate with rust_analyzer it would be nice.

**Comment by l1nxy** [comment]
I don't think it is needed, there has not any lang server provide such feature can make the lang server become to "meta" lang server, so r-a  does not to do.

**Comment by glyh** [comment]
There wasn't any real language that had rust's lifetime borrow checking before rust, so rust doesn't need to exist.
